src/dag/gc.test.ts

```diff
@@ -0,0 +1,227 @@
+import {expect} from '@esm-bundle/chai';
+import {fakeHash, Hash} from '../hash';
+import {computeRefCountUpdates, GarbageCollectionDelegate} from './gc';
+
+function createGraph(args: {
+  graph: Record<string, string[]>;
+  heads: string[];
+  allZeroRefCounts?: boolean;
+}) {
+  const {graph, heads, allZeroRefCounts} = args;
+  const hashes = Object.fromEntries(
+    Object.keys(graph).map(k => [k.toString(), fakeHash(k)]),
+  );
+  const refs = Object.fromEntries(
+    Object.entries(graph).map(([k, refs]) => [
+      hashes[k].toString(),
+      refs.map(v => hashes[v]),
+    ]),
+  );
+
+  const refCounts = Object.fromEntries(
+    Object.keys(graph).map(k => [hashes[k].toString(), 0]),
+  );
+
+  if (!allZeroRefCounts) {
+    const q = Array.from(new Set(heads));
+    for (const k of q) {
+      refCounts[hashes[k].toString()] = refCounts[hashes[k].toString()] + 1;
+      q.push(...graph[k]);
+    }
+  }
+
+  const delegate: GarbageCollectionDelegate = {
+    getRefCount: async hash => refCounts[hash.toString()] || 0,
+    getRefs: async hash => refs[hash.toString()] || [],
+  };
+
+  return {
+    hashes,
+    refCounts,
+    refs,
+    delegate,
+  };
+}
+
+function expectRefCountUpdates(
+  actual: Map<Hash, number>,
+  expected: Record<string, number>,
+) {
+  const expectedAsMap = new Map(
+    Object.entries(expected).map(([k, v]) => [fakeHash(k), v]),
+  );
+  expect(actual).to.deep.equal(expectedAsMap);
+}
+
+test('computeRefCountUpdates includes entry for every putChunk', async () => {
+  //   R    C
+  //  / \   |
+  // A   B  D
+  const {hashes, delegate} = createGraph({
+    graph: {
+      r: ['a', 'b'],
+      a: [],
+      b: [],
+      c: ['d'],
+      d: [],
+    },
+    heads: [],
+    allZeroRefCounts: true,
+  });
+
+  const refCountUpdates = await computeRefCountUpdates(
+    [{old: undefined, new: hashes['r']}],
+    new Set(Object.values(hashes)),
+    delegate,
+  );
+  expectRefCountUpdates(refCountUpdates, {
+    r: 1,
+    a: 1,
+    b: 1,
+    c: 0,
+    d: 0,
+  });
```

Comment:
I think this test would be easier to read if it tested the current ref count value and not only the ref count updates